### PR TITLE
ARM Template in Bicep

### PR DESCRIPTION
**Pull request by midhun6989**
1. Added ARM Template in Bicep
2. Modified ARM Template in JSON to match with the ARM Template in Bicep
3. Modified GitHub Action workflow file to do the OCP Cluster deployment using the ARM Template in Bicep

Related to the issue https://github.com/IBM/zmodstack-deploy/issues/18.

### CODE REVIEW

File: .github/workflows/azure-deploy.yml [hunk 1]
```diff
@@ -16,33 +16,44 @@ on:
...
       uses: azure/arm-deploy@v1
       with:
+        scope: resourcegroup
+        region: ${{ env.RESOURCE_GROUP_LOCATION }}
         resourceGroupName: ${{ env.RESOURCE_GROUP_NAME }}
-        template: ${{ env.ARM_TEMPLATE }}
-        parameters: ${{ env.ARM_PARAMETERS }}
+        deploymentName: ${{ env.DEPLOYMENT_NAME }}
+        template: ${{ env.ARM_TEMPLATE_BICEP }}
+        parameters: ${{ env.ARM_PARAMETERS }}
```
Comment (ivandov) [inline .github/workflows/azure-deploy.yml:59]: Aren't these steps performed in the Ansible playbook you created as well? I was imagining that the Github action steps here should just define the order of execution, but delegate the functional steps to the Ansible playbook layer. 

Not a PR blocker, but something to consider for refactoring later.

**Review by ivandov — APPROVED** [review]
Changes look good to me. Added some comments for future consideration. 

Also, it looks like the generated `mainTemplate.json` will have some metadata inserted that's new for every build? Will that cause any confusion with git commit history on this file? Should we move the generated JSON to be something that we hold as an asset _outside_ of the git commits?

**Comment by midhun6989** [comment]
> Changes look good to me. Added some comments for future consideration.
> 
> Also, it looks like the generated `mainTemplate.json` will have some metadata inserted that's new for every build? Will that cause any confusion with git commit history on this file? Should we move the generated JSON to be something that we hold as an asset _outside_ of the git commits?

@ivandov I have regenerated the `mainTemplate.json` in the GitHub repository to match with the `mainTemplate.bicep`. What I am thinking is to update the `mainTemplate.json` whenever there are some changes in `mainTemplate.bicep`, so that both files will be in sync. Also, in the GitHub Action, for ARM-TTK execution, there is a similar `mainTemplate.json` created from `mainTemplate.bicep`. Not sure if I have answered your query.

File: .github/workflows/azure-deploy.yml (hunk 1 above)
Comment (midhun6989) [reply .github/workflows/azure-deploy.yml:59]: > Aren't these steps performed in the Ansible playbook you created as well? I was imagining that the Github action steps here should just define the order of execution, but delegate the functional steps to the Ansible playbook layer.
> 
> Not a PR blocker, but something to consider for refactoring later.

@ivandov The order of execution is GitHub Action ---> ARM Template ---> Bootstrap Script ---> Ansible Playbook.
One of the prerequisites for executing `azure/arm-deploy@v1` in GitHub Action is the availability of Resource Group beforehand. Thus a Resource Group is created  explicitly. Followed the documentation at https://github.com/Azure/arm-deploy#another-example-which-ensures-the-azure-resource-group-exists-before-arm-deployment. Hope I have answered your query.

**Pull request merged**